src/slot.rs

```diff
@@ -642,102 +648,238 @@ impl Stats {
+            self.pool_name.as_str(),
+            id.0,
+            handle.slot_id().as_u64()
+        ));
+
+        return Ok(handle);
     }
 
-    // Creates a new Slot, which always starts as "Connecting", and spawn a task
-    // to actually connect to the backend and monitor slot health.
+    // Borrows a connection out of the first unclaimed slot.
+    //
+    // Returns a Handle which has enough context to put the claim back,
+    // once it's dropped by the client.
     #[instrument(
-        skip(self)
+        skip(self, permit)
         fields(pool_name = %self.pool_name),
-        name = "SetWorker::create_slot"
+        name = "Slots::take_connected_unclaimed_slot"
     )]
+    fn take_connected_unclaimed_slot(
+        &mut self,
+        permit: mpsc::OwnedPermit<BorrowedConnection<Conn>>,
+        claim_id: ClaimId,
+    ) -> Option<claim::Handle<Conn>> {
+        for (id, slot) in &self.slots {
+            let guarded = slot.inner.guarded.lock().unwrap();
+            event!(Level::TRACE, slot_id = id.as_u64(), state = ?guarded.state, "Considering slot");
+            if matches!(guarded.state, State::ConnectedUnclaimed(_)) {
+                event!(Level::TRACE, slot_id = id.as_u64(), "Found unclaimed slot");
+                // We intentionally "take the connection out" of the slot and
+                // "place it into a claim::Handle" in the same method.
+                //
+                // This makes it difficult to leak a connection, unless the drop
+                // method of the claim::Handle is skipped.
+                let State::ConnectedUnclaimed(DebugIgnore(conn)) = slot.inner.state_transition(
+                    *id,
+                    &self.backend,
+                    guarded,
+                    State::ConnectedClaimed,
+                ) else {
+                    panic!(
+                        "We just matched this type before replacing it; this should be impossible"
+                    );
+                };
+
+                let borrowed_conn = BorrowedConnection::new(conn, *id);
+
+                #[cfg(feature = "probes")]
+                crate::probes::handle__claimed!(|| {
+                    (
+                        self.pool_name.as_str(),
+                        claim_id.0,
+                        borrowed_conn.id.as_u64(),
+                        self.backend.address.to_string(),
+                    )
+                });
+
+                // The "drop" method of the claim::Handle will return it to
+                // the slot set, through the permit (which is connected to
+                // slot_rx).
+                return Some(claim::Handle::new(borrowed_conn, permit));
+            }
+        }
+        None
+    }
+
+    fn set_wanted_count(&mut self, count: usize) {
+        self.wanted_count = std::cmp::min(count, self.config.max_count);
+        self.conform_slot_count();
+    }
+
+    fn conform_slot_count(&mut self) {
```

Comment:
This function has basically been moved from the `SetWorker` to this new struct, `Slots`